src/coreclr/gc/gc.cpp

```diff
@@ -16,6 +16,43 @@
 // allocation helpers in gcscan.cpp
 //
 
+#include "common.h"
+#include "gcenv.h"
+
+#include "gc.h"
+#include "gcscan.h"
+#include "gcdesc.h"
+#include "softwarewritewatch.h"
+#include "handletable.h"
+#include "handletable.inl"
+#include "gcenv.inl"
+#include "gceventstatus.h"
+
+#ifdef __INTELLISENSE__
+#if defined(FEATURE_SVR_GC)
+
+#define SERVER_GC 1
+
+#else // defined(FEATURE_SVR_GC)
+
+#ifdef SERVER_GC
+#undef SERVER_GC
+#endif
+
+#endif // defined(FEATURE_SVR_GC)
+#endif // __INTELLISENSE__
```

Comment:
I don't think I understand how intellisense works.  How does this help?  In a typical build, FEATURE_SVR_GC is defined; I believe that it indicates whether to build SVR in addition to WKS, not which one is currently being built (SERVER_GC and then mostly MULTIPLE_HEAPS handle that).

Reply:
We have separate targets for NativeAOT. In this case we can select either `Runtime.ServerGC.lib` or `Runtime.WorkStationGC.lib` in Visual Studio so that only one of them will be highlighted, while another one will be faded out.
![image](https://github.com/user-attachments/assets/aac21648-f39e-428a-9c9b-abbe55da7167)
This is helpful if we are only working on a particular GC.

Reply:
I think I see.. so it's something like this?
- This change doesn't do anything to improve intellisense's handling of gc.cpp being included twice in a project
- This change does set up gc.cpp to be self-contained so that for each project intellisense will yield one correct/consistent view of gc.cpp.
- Restating the above, non-AOT builds don't get the two different intellisense views of gc.cpp.  However, likely now get a more consistent view of SVR GC.
- Non-AOT development can theoretically use these two AOT builds as views into the two GCs with the caveat to be careful around NATIVEAOT ifdefs.

Are these two AOT libs used for shipping or purely for intellisense?  Do you know why SERVER_GC_SOURCES is used to include/exclude gcsvr.cpp when FEATURE_SVR_GC is enough to render it empty?  Or why (due to COMMON_RUNTIME_SOURCES) Runtime.ServerGC.lib seems to include WKS GC rather than making gcwks.cpp opt-in (or adding a FEATURE_WKS_GC)?

Reply:
> This change doesn't do anything to improve intellisense's handling of gc.cpp being included twice in a project

Yes. Before the change, if you open gc.cpp directly, the intellisense engine doesn't know the gc.cpp is put under a namespace, as well as those include files in its parent cpp files, so it yields a lot of errors. 

> This change does set up gc.cpp to be self-contained so that for each project intellisense will yield one correct/consistent view of gc.cpp.

Yes. By making gc.cpp self-contained now the intellisense engine has all necessary information to function properly. 

> Restating the above, non-AOT builds don't get the two different intellisense views of gc.cpp. However, likely now get a more consistent view of SVR GC.

Yes, non-AOT builds can only see SVC GC is activated in the intellisense (except the `gcexample.exe` one which is using WKS GC). 

> Non-AOT development can theoretically use these two AOT builds as views into the two GCs with the caveat to be careful around NATIVEAOT ifdefs.

Yes, if we create separate cmake targets for them, see below. 

> Are these two AOT libs used for shipping or purely for intellisense?

They are shipping together with the NativeAOT runtime. Those targets are defined in https://github.com/dotnet/runtime/blob/9fe8c95c890b0e68c77c7970f8413b1375afe18f/src/coreclr/nativeaot/Runtime/Full/CMakeLists.txt#L29-L33.

Reply:
Thanks for the detailed replies.  I think that I understand all of this except for the SVR vs WKS build.  I understand the goal of including one GC in a binary but not the mechanisms:
- The runtime is already set up to include/exclude various features at the build-level with the FEATURE_* macros.  FEATURE_SVR_GC is one of them; if it is not set, then the entire gcsvr.cpp file is empty (except, technically, for a comment and including common.h).  So I still don't understand why the nativeaot build uses a different mechanism to exclude SVR GC.
- Runtime.ServerGC.lib is choosing to include SVR GC, but I don't see anything that can exclude WKS GC (whether that is excluding gcwks.cpp or preferably adding a FEATURE_WKS_GC like other features).  Is Runtime.ServerGC.lib much bigger than Runtime.WorkstationGC.lib?

Reply:
SVR GC falls back to WKS GC for single-proc machines. We do not have a pure SVR GC config since such fallback would not be possible in such config.

Reply:
Oops, yeah. I just realized that we didn't build an SVR-only GC. 
Then I think using `FEATURE_SVR_GC` should be enough here. 
@MichalStrehovsky do you know why we are including gc cpp files manually for NativeAOT libs instead of using `FEATURE_SVR_GC`?

Reply:
The build setup is the same for regular CoreCLR and NativeAOT - both include gc cpp files manually:

- Regular CoreCLR: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/CMakeLists.txt#L482
- NativeAOT: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/Runtime/CMakeLists.txt#L59

I do not think there is particular reason for why it needs to be done this way. There are certainly other ways possible to setup the build that would produce the same result.

The only special thing about native AOT is that it builds WKS-only flavor. There is no WKS-only flavor for regular CoreCLR. Regular CoreCLR always includes both WKS and SVR.